### PR TITLE
Update Cipher documentation.

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/man7/provider-cipher.pod [hunk 1]
```diff
@@ -202,272 +202,9 @@ with the provider side context I<cctx> in its current state if it is
...
-
-=back
-
-The default is "CS1".
-This is only supported for "AES-128-CBC-CTS", "AES-192-CBC-CTS" and "AES-256-CBC-CTS".
-
-=back
+Parameters currently recognised by built-in ciphers are listed in
+L<EVP_EncryptInit(3)/PARAMETERS>.
+Not all parameters are relevant to, or are understood by all ciphers:
```
Comment (paulidale) [inline doc/man7/provider-cipher.pod:207]: Full stop not colon.

File: providers/implementations/ciphers/cipher_aes_siv.c [hunk 2]
```diff
@@ -185,7 +185,6 @@ static int aes_siv_get_ctx_params(void *vctx, OSSL_PARAM params[])
 static const OSSL_PARAM aes_siv_known_gettable_ctx_params[] = {
     OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_KEYLEN, NULL),
     OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_AEAD_TAGLEN, NULL),
-    OSSL_PARAM_uint(OSSL_CIPHER_PARAM_SPEED, NULL),
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_siv.c:188]: This is noted as supported by SIV in the documentation, why remove it ?

**Review by paulidale — CHANGES_REQUESTED** [review]
Looks good.

File: doc/man3/EVP_EncryptInit.pod [hunk 3]
```diff
@@ -197,70 +197,132 @@ symmetric ciphers.
...
+that can be used with EVP_CIPHER_CTX_set_params().
+EVP_CIPHER_settable_ctx_params() returns the parameters that can be set from the
+algorithm, whereas EVP_CIPHER_CTX_settable_params() returns the parameters that
+can be set in the context's current state.
+See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as a parameter descriptor.
+
+=item EVP_EncryptInit_ex2()
+
+Sets up cipher context I<ctx> for encryption with cipher I<type>. I<type> is
+typically supplied by a function such as EVP_aes_256_cbc(), or a value
```
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod:293]: Both are valid.

File: doc/man3/EVP_EncryptInit.pod [hunk 4]
```diff
@@ -197,70 +197,132 @@ symmetric ciphers.
...
-specified.
+=item EVP_EncryptInit_ex()
 
-EVP_EncryptUpdate() encrypts B<inl> bytes from the buffer B<in> and
-writes the encrypted version to B<out>. This function can be called
-multiple times to encrypt successive blocks of data. The amount
-of data written depends on the block alignment of the encrypted data.
+Is a legacy function similiar to EVP_EncryptInit_ex2() except If I<impl>
+is non-NULL, its implementation of the cipher B<type> is used if there is one,
+and if not, the default implementation is used.
```
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod:309]: What about

This legacy function is similar to EVP_EncryptInit_ex2() when I<impl> is NULL.
The I<cipher> field will be replaced by the engines I<impl> internal cipher if
it exists.

File: doc/man3/EVP_EncryptInit.pod (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply doc/man3/EVP_EncryptInit.pod:293]: One isn't recommended.  It's the first one listed and the easier/lazier to use.  Move it later to a separate sentence perhaps?

File: doc/man3/EVP_EncryptInit.pod [hunk 5]
```diff
@@ -276,180 +338,182 @@ If padding is disabled then EVP_EncryptFinal_ex() will not encrypt any more
...
+These functions that can be used for decryption or encryption. The operation
+performed depends on the value of the I<enc> parameter. It should be set to 1
+for encryption, 0 for decryption and -1 to leave the value unchanged
+(the actual value of 'enc' being supplied in a previous call).
+
+=item EVP_CIPHER_CTX_reset()
+
+Clears all information from a cipher context and free up any allocated memory
+associated with it, except the I<ctx> itself. This function should be called
+anytime I<ctx> is to be reused for another
```
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod:364]: Changed to 'is reused by another'

File: doc/man3/EVP_EncryptInit.pod (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply doc/man3/EVP_EncryptInit.pod:309]: :+1:

File: doc/man3/EVP_EncryptInit.pod [hunk 6]
```diff
@@ -276,180 +338,182 @@ If padding is disabled then EVP_EncryptFinal_ex() will not encrypt any more
...
+
+Return the block size of a cipher when passed an B<EVP_CIPHER> or
+B<EVP_CIPHER_CTX> structure. The constant B<EVP_MAX_BLOCK_LENGTH> is also the
+maximum block length for all ciphers.
+
+=item EVP_CIPHER_type() and EVP_CIPHER_CTX_type()
+
+Return the type of the passed cipher or context. This "type" is the actual NID
+of the cipher OBJECT IDENTIFIER as such it ignores the cipher parameters and
+40 bit RC2 and 128 bit RC2 have the same NID. If the cipher does not have an
```
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod:447]: Put in ()

File: doc/man3/EVP_EncryptInit.pod [hunk 7]
```diff
@@ -276,180 +338,182 @@ If padding is disabled then EVP_EncryptFinal_ex() will not encrypt any more
...
+
+Returns the internal dynamic number assigned to the I<cipher>.  This is only
+useful with fetched B<EVP_CIPHER>s.
+
+=item EVP_CIPHER_name() and EVP_CIPHER_CTX_name()
 
-EVP_CIPHER_number() returns the internal dynamic number assigned to
-the I<cipher>.  This is only useful with fetched B<EVP_CIPHER>s.
+Return the name of the passed cipher or context.  For fetched ciphers with
+multiple names, only one of them is returned; it's recommended to use
```
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod:467]: CHanged to 'See also EVP_CIPHER_names_do_all().'

File: doc/man3/EVP_EncryptInit.pod [hunk 8]
```diff
@@ -458,18 +522,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+=item "iv" (B<OSSL_CIPHER_PARAM_IV>) <octet string OR octet ptr>
+
+Gets the IV used to initialize the associated cipher context I<ctx>.
+See also EVP_CIPHER_CTX_get_original_iv().
+
+=item "updated-iv" (B<OSSL_CIPHER_PARAM_UPDATED_IV>) <octet string OR octet ptr>
+
+Gets the updated pseudo-IV state for the associated cipher ctx, e.g.,
+the previous ciphertext block for CBC mode or the iteratively encrypted IV
+value for OFB mode.  Note that octet pointer access is deprecated and is
```
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod:720]: We have an old API that needs this functionality.. So dont use it is what this is trying to say.

File: providers/implementations/ciphers/cipher_aes_siv.c (hunk 2 above)
Comment (slontis) [reply providers/implementations/ciphers/cipher_aes_siv.c:188]: Becuase this has only ever been a settable control. And it is not implemented by the get_ctx_params() (And doesnt need to be).

File: doc/man3/EVP_EncryptInit.pod (hunk 8 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply doc/man3/EVP_EncryptInit.pod:719]: Sigh.

File: doc/man3/EVP_EncryptInit.pod [hunk 9]
```diff
@@ -197,70 +197,132 @@ symmetric ciphers.
...
 
-EVP_CIPHER_fetch() fetches the cipher implementation for the given
-B<algorithm> from any provider offering it, within the criteria given
-by the B<properties>.
+=over 4
+
+=item EVP_CIPHER_fetch()
+
+Fetches the cipher implementation for the given B<algorithm> from any provider
+offering it, within the criteria given by the B<properties>.
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:205]: Could you please replace the occurrences of `B<argument>` with `I<argument>`? It seems the file is wildly inconsistent in regards to this.

File: doc/man3/EVP_EncryptInit.pod [hunk 10]
```diff
@@ -197,70 +197,132 @@ symmetric ciphers.
...
+=item EVP_CIPHER_CTX_free()
+
+Clears all information from a cipher context and frees any allocated memory
+associated with it, including B<ctx> itself. This function should be called after
+all operations using a cipher are complete so sensitive information does not
+remain in memory.
+
+=item EVP_CIPHER_CTX_ctrl()
+
+I<This is a legacy method. EVP_CIPHER_CTX_set_params() and
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:234]: Not sure how the rendered markup will look like with function names inside `I<>`. Maybe put only the first sentence inside it?

File: doc/man3/EVP_EncryptInit.pod [hunk 11]
```diff
@@ -197,70 +197,132 @@ symmetric ciphers.
...
 specified.
 
-EVP_EncryptUpdate() encrypts B<inl> bytes from the buffer B<in> and
-writes the encrypted version to B<out>. This function can be called
-multiple times to encrypt successive blocks of data. The amount
-of data written depends on the block alignment of the encrypted data.
+=item EVP_EncryptInit_ex()
+
+This legacy function is similar to EVP_EncryptInit_ex2() when I<impl> is NULL.
+The I<cipher> field will be replaced by the engines I<impl> internal cipher if
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:308]: This is a slightly confusing sentence as it refers to a parameter named differently in the prototype. I'd use something like `The implementation of the I<type> from the I<impl> engine will be used if it exists.`

File: doc/man3/EVP_EncryptInit.pod [hunk 12]
```diff
@@ -276,180 +338,181 @@ If padding is disabled then EVP_EncryptFinal_ex() will not encrypt any more
...
+
+Clears all information from a cipher context and free up any allocated memory
+associated with it, except the I<ctx> itself. This function should be called
+anytime I<ctx> is reused by another
+EVP_CipherInit() / EVP_CipherUpdate() / EVP_CipherFinal() series of calls.
+
+=item EVP_EncryptInit(), EVP_DecryptInit() and EVP_CipherInit()
+
+Behave in a similar way to EVP_EncryptInit_ex(), EVP_DecryptInit_ex() and
 EVP_CipherInit_ex() except they always use the default cipher implementation.
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:370]: I'd suggest: `... except if the I<type> is not a fetched cipher they use the default implementation of the I<type>.`

File: doc/man3/EVP_EncryptInit.pod [hunk 13]
```diff
@@ -276,180 +338,181 @@ If padding is disabled then EVP_EncryptFinal_ex() will not encrypt any more
...
+=item EVP_Cipher()
+
+Encrypts or decrypts a maximum I<inl> amount of bytes from I<in> and leaves the
+result in I<out>.
 If the cipher doesn't have the flag B<EVP_CIPH_FLAG_CUSTOM_CIPHER> set,
 then I<inl> must be a multiple of EVP_CIPHER_block_size().  If it isn't,
 the result is undefined.  If the cipher has that flag set, then I<inl>
 can be any size.
 This function is historic and shouldn't be used in an application, please
 consider using EVP_CipherUpdate() and EVP_CipherFinal_ex instead.
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:388]: Not sure this sentence is needed. If it was that much historic we should probably deprecate it, but I do not see us doing it for 3.0. What about: `Due to the constraints of the API contract of this function it shouldn't be used in applications, please consider using EVP_CipherUpdate() and EVP_CipherFinal_ex instead.`

File: doc/man3/EVP_EncryptInit.pod [hunk 14]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
 or the parameters cannot be set (for example the RC2 effective key length
 is not supported.
 
-EVP_CIPHER_CTX_ctrl() allows various cipher specific parameters to be determined
-and set.
+=item EVP_CIPHER_CTX_rand_key()
+
+Generates a random key of the appropriate length based on the cipher context.
+The B<EVP_CIPHER> can provide its own random key generation routine to support
+keys of a specific form. I<Key> must point to a buffer at least as big as the
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:528]: `I<key>`

File: doc/man3/EVP_EncryptInit.pod [hunk 15]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+
+Gets the updated pseudo-IV state for the associated cipher ctx, e.g.,
+the previous ciphertext block for CBC mode or the iteratively encrypted IV
+value for OFB mode.  Note that octet pointer access is deprecated and is
+provided only for backwards compatibility with historical libcrypto APIs.
+See also EVP_CIPHER_CTX_get_updated_iv().
+
+=item "randkey" (B<OSSL_CIPHER_PARAM_RANDOM_KEY>) <octet string>
+
+Gets a implementation specific randomly generated key for the associated
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:725]: Nit: an implementation

File: doc/man3/EVP_EncryptInit.pod [hunk 16]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+cipher ctx I<ctx>. This is currently only supported by DES and 3DES (which set
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:726]: Nit: `cipher context` otherwise it looks too silly.

File: doc/man3/EVP_EncryptInit.pod [hunk 17]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+the key to odd parity).
+
+=item "taglen" (B<OSSL_CIPHER_PARAM_AEAD_TAGLEN>) <unsigned integer>
+
+Gets the tag length to be used for an AEAD cipher for the associated cipher ctx
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:731]: as above cipher context

File: doc/man3/EVP_EncryptInit.pod [hunk 18]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+I<ctx>. It gets a default value if it has not been set.
+The length of the "taglen" parameter should not exceed that of a B<size_t>.
+See also EVP_CIPHER_CTX_tag_length().
+
+=item "tlsaadpad" (B<OSSL_CIPHER_PARAM_AEAD_TLS1_AAD_PAD>) <unsigned integer>
+
+Gets the length of the tag that will be added to a TLS record for the AEAD
+tag for the associated cipher ctx I<ctx>.
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:739]: same as above or you could also shortcut it to: `tag for the associated I<ctx>.`

File: doc/man3/EVP_EncryptInit.pod [hunk 19]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+
+=item "tls-mac" (B<OSSL_CIPHER_PARAM_TLS_MAC>) <octet ptr>
+
+Used to pass the TLS MAC data.
+
+=back
+
+=head2 Settable EVP_CIPHER_CTX parameters
+
+The following OSSL_PARAM keys can be used with EVP_CIPHER_CTX_set_params():
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:769]: `B<OSSL_PARAM>`?

File: doc/man3/EVP_EncryptInit.pod [hunk 20]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+=item "tls1multi_interleave" (B<OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_INTERLEAVE>) <unsigned integer>
+
+Sets or gets the number of records being sent in one go for a tls1 multiblock
+cipher operation (either 4 or 8 records).
+
+=back
+
+=head2 Gettable EVP_CIPHER_CTX parameters
+
+The following OSSL_PARAM keys can be used with EVP_CIPHER_CTX_get_params():
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:700]: `B<OSSL_PARAM>`?

File: doc/man3/EVP_EncryptInit.pod [hunk 21]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+otherwise it gets 0. The interleaving is an optimization only applicable to certain
+TLS ciphers.
+Use (EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_TLS1_1_MULTIBLOCK) to retrieve the
+cached value.
+
+=back
+
+=head2 Gettable and Settable EVP_CIPHER_CTX parameters
+
+The following OSSL_PARAM keys can be used with both EVP_CIPHER_CTX_get_params()
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:613]: `B<OSSL_PARAM>`

File: doc/man3/EVP_EncryptInit.pod [hunk 22]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+
+=over 4
+
+=item "mackey" (B<OSSL_CIPHER_PARAM_AEAD_MAC_KEY>) <octet string>
+
+Sets the MAC key used by composite AEAD ciphers such as AES-CBC-HMAC-SHA256.
+
+=item "speed" (B<OSSL_CIPHER_PARAM_SPEED>) <unsigned integer>
+
+Sets the speed option for the associated cipher ctx. This is only supported
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:779]: context

File: doc/man3/EVP_EncryptInit.pod [hunk 23]
```diff
@@ -458,18 +521,472 @@ possible for this function to fail if the cipher does not have any ASN1 support
...
+
+Sets the TLS version.
+
+=item "tls-mac-size" (B<OSSL_CIPHER_PARAM_TLS_MAC_SIZE>) <unsigned integer>
+
+Set the TLS MAC size.
+
+=item "tlsaad" (B<OSSL_CIPHER_PARAM_AEAD_TLS1_AAD>) <octet string>
+
+Sets TLSv1.2 AAD information for the associated cipher ctx I<ctx>.
```
Comment (t8m) [inline doc/man3/EVP_EncryptInit.pod:794]: As above

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/man3/EVP_EncryptInit.pod (hunk 3 above)
[the remark this answers is not in the document]
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod]: Put in a separate sentence

File: doc/man3/EVP_EncryptInit.pod (hunk 9 above)
Comment (slontis) [reply doc/man3/EVP_EncryptInit.pod:205]: Done - that is what happens when you block copy the contents of one file to another :)

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. Thank you.

**Pull request closed without merging**